service/src/lib.rs

```diff
@@ -39,59 +35,50 @@ pub trait Factory: Send + Sync {
     fn get_metadata(&self) -> DeploymentMetadata;
 }
 
-/// Used to get resources of type `T` from factories.
-///
-/// This is mainly meant for consumption by our code generator and should generally not be called by users.
-///
-/// ## Creating your own managed resource
-///
-/// You may want to create your own managed resource by implementing this trait for some builder `B` to construct resource `T`.
-/// [`Factory`] can be used to provision resources on Shuttle's servers if your service will need any.
-///
-/// Please refer to `shuttle-examples/custom-resource` for examples of how to create custom resource. For more advanced provisioning
-/// of custom resources, please [get in touch](https://discord.gg/shuttle) and detail your use case. We'll be interested to see what you
-/// want to provision and how to do it on your behalf on the fly.
-///
-/// ```
+/// Allows implementing plugins for the Shuttle main function.
 #[async_trait]
-pub trait ResourceBuilder<T> {
-    /// The type of resource this creates
-    const TYPE: Type;
+pub trait ResourceBuilder {
+    /// The type of resource this plugin creates.
+    const TYPE: resource::Type;
 
-    /// The internal config being constructed by this builder. This will be used to find cached [Self::Output].
-    type Config: Serialize;
+    /// The input config to this resource.
+    type Config: Default + Serialize;
 
-    /// The output type used to build this resource later
+    /// The output from requesting this resource.
+    /// A cached copy of this will be used if the same [`ResourceBuilder::Config`] is found for this [`ResourceBuilder::TYPE`].
     type Output: Serialize + DeserializeOwned;
```

Comment:
I'm still thinking `Output` might want `IntoResource<R>` bound as well. But that would add generic `<R>` back to `ResourceBuilder` and defeat the flexibility that we are adding here (one `ResourceBuilder<R>` impl per final output type?).

Reply:
Hmm, yeah that will break the flexibility.

Reply:
Some `Output` might not need `IntoResource` to other types. Maybe we should have a default `impl<T> IntoResource<T> for T`?